build: Upload source maps to Sentry

dcalhoun opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/6613.

## Proposed Changes

- Upload source maps for error reporting using the `@sentry/webpack-plugin`
package.
- Enable environment configuration files via the `dotenv` package. 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Uploaded source maps are displayed in error reports**

> [!IMPORTANT]
> Some error reports originating from the _main process_ lack stack traces entirely, which results in the uploaded source maps having no effect. It is unknown at this time why some error reports lack stack traces. It may relate to whether the `event.processor` is `browser` or `renderer`. 

1. Copy `.env.example` to `.env`.
1. Populate the `.env` Sentry value with a valid authorization token. A new "User Auth Token" can be used with the  `project:read, project:releases, project:write` scopes. 
1. Introduce an intentional error (example patch below).
1. Run `npm run make`.
1. Launch the app found in `./out`.
1. Trigger the crash.
1. Verify the error report includes a stack trace referencing the source files,
   not minified files.

<details><summary>Intentional error diff</summary>

Clicking to copy a value in a site's _Settings_ pane will throw an error: 

```diff
diff --git a/src/components/copy-text-button.tsx b/src/components/copy-text-button.tsx
index fb8091e..3ff7082 100644
--- a/src/components/copy-text-button.tsx
+++ b/src/components/copy-text-button.tsx
@@ -27,12 +27,13 @@ export function CopyTextButton( {
 	const [ timeoutId, setTimeoutId ] = useState< NodeJS.Timeout | null >( null );
 
 	const onClick = useCallback( () => {
-		getIpcApi().copyText( text );
-		setShowCopied( true );
-		if ( timeoutId ) {
-			clearTimeout( timeoutId );
-		}
-		setTimeoutId( setTimeout( () => setShowCopied( false ), timeoutConfirmation ) );
+		throw new Error( 'An intentional testing error' );
+		// getIpcApi().copyText( text );
+		// setShowCopied( true );
+		// if ( timeoutId ) {
+		// 	clearTimeout( timeoutId );
+		// }
+		// setTimeoutId( setTimeout( () => setShowCopied( false ), timeoutConfirmation ) );
 	}, [ text, timeoutConfirmation, timeoutId ] );
 
 	return (

```

</details>

**Environment configuration values are not within bundled files**

> [!NOTE]
> This example is a bit contrived, but done purely to showcase environment variables are not included in the compiled source by default. 

> [!IMPORTANT]
> The provided example `.env` file values will cause `npm run make` to throw errors due to the lack of an authentication token. However, the build should still proceed and complete successfully, producing compiled JS files in the `.webpack` directory. 

1. Copy `.env.example` to `.env`.
1. Populate the `.env` file two a arbitrary values and reference on of the two
   values in the source (example patch below).
1. Run `npm run make`.
1. Verify the resulting source does not include value that are not explicitly
   configured for the bundle.

<details><summary>Excluded values example</summary>

**Example `.env` file testing values:** 
```shell
FIRST_VAR="a-unique-value-01"
SECOND_VAR="a-unique-value-02"

```

**Example variable reference diff:** 
```diff
diff --git a/src/index.ts b/src/index.ts
index 96e790c..93b6b7e 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -25,6 +25,12 @@ import setupWPServerFiles from './setup-wp-server-files';
 import { setupUpdates } from './updates';
 import { stopAllServersOnQuit } from './site-server'; // eslint-disable-line import/order
 
+// After building the JS, performing a project-wide search of the compiled JS
+// files in the `.webpack` directory for the `a-unique-value-01` value should
+// locate this line. Contrastingly, searching for the `a-unique-value-02` value
+// should not locate _any_ lines in the project.
+console.log( '>>>', process.env.FIRST_VAR );
+
 Sentry.init( {
 	dsn: 'https://97693275b2716fb95048c6d12f4318cf@o248881.ingest.sentry.io/4506612776501248',
 	debug: true,
diff --git a/webpack.plugins.ts b/webpack.plugins.ts
index 4a7bd23..970ece0 100644
--- a/webpack.plugins.ts
+++ b/webpack.plugins.ts
@@ -1,4 +1,5 @@
 import { sentryWebpackPlugin } from '@sentry/webpack-plugin';
+import { EnvironmentPlugin } from 'webpack';
 import type IForkTsCheckerWebpackPlugin from 'fork-ts-checker-webpack-plugin';
 import type { WebpackPluginInstance } from 'webpack';
 import 'dotenv/config';
@@ -13,6 +14,10 @@ export const plugins: WebpackPluginInstance[] = [
 			},
 		},
 	} ),
+	new EnvironmentPlugin( {
+		// Only the first of two variables are provided to the source
+		FIRST_VAR: process.env.FIRST_VAR,
+	} ),
 	// Sentry must be the last plugin
 	sentryWebpackPlugin( {
 		authToken: process.env.SENTRY_AUTH_TOKEN,

```

</details>

| Before | After |
| - | - |
| ![example-stack-trace-before](https://github.com/Automattic/studio/assets/438664/401ea6c8-bdaa-4d9f-952e-a100035d7934) | ![example-stack-trace-after](https://github.com/Automattic/studio/assets/438664/7dd64b7b-3c92-4ca4-bb4c-f2bc96929884) |

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Have you checked for TypeScript, React or other console errors?

dcalhoun committed:
build: Upload source maps to Sentry

Improve stack traces provided in Sentry issues.

dcalhoun committed:
build: Create example environment configuration file

dcalhoun committed:
build: Add dotenv package

Enable environment configuration file.

dcalhoun committed:
build: Load environment configuration files for Webpack build

E.g., load necessary Sentry authentication token for the Webpack build
process to upload source maps.

dcalhoun committed:
refactor: Co-locate environment configuration file ignores

Improve comprehensibility by grouping related logic.

dcalhoun commented on .gitignore:
```diff
@@ -61,6 +61,9 @@ typings/
 .env
 .env.test
 
+# Sentry Config File
+.env.sentry-build-plugin
```
This file is automatically supported by the `@sentry/webpack-plugin`, so I left this ignore configuration generated by the setup wizard in place.

sejas reviewed (approved):
It works great! Thanks for making the errors more clear 🙏 .

I confirm the source code is available in Sentry after raising an error on the Renderer, but no in NodeJS.

I've created a new token from sentry `/settings/account/api/auth-tokens/new-token/`
Introduced a `throw new Error...`
Run `npm run make`
Reproduce your error
Go to Sentry, filter by 1h and verify the sourcecode is there.
Remove your token

My issues: `/issues/5278414006`, and `/issues/5278367471`

<img width="672" alt="Screenshot 2024-04-29 at 18 54 08" src="https://github.com/Automattic/studio/assets/779993/9af54d4c-df59-477e-84f5-b79957e4a02a">

sejas commented:
Note that I couldn't see the source code when the error came from NodeJS, like `/issues/5278456660/`.

dcalhoun commented:
> Note that I couldn't see the source code when the error came from NodeJS, like `/issues/5278456660/`.

I spoke to @fluiddot about this earlier today. Our current belief is that [Electron's limitation](https://www.electronjs.org/docs/latest/api/ipc-main#ipcmainhandlechannel-listener) regarding errors communicated over the IPC (quoted below) is what causes the lack of stack trace for these types of errors. 

> Errors thrown through handle in the main process are not transparent as they are serialized and only the message property from the original error is provided to the renderer process. Please refer to [#24427](https://github.com/electron/electron/issues/24427) for details.

Moving forward, we might refactor our IPC methods so that they report errors to Sentry prior to re-throwing them for the renderer. That would also us to both report a stack trace and display a helpful message for the user.

mokagio commented:
@dcalhoun is there an expectations on local builds to upload source maps?

The fact that the testing instructions and the code expect a local `.env` makes me think there might be, but I'm not sure whether that's _necessary_. In my experience on other platforms, tracking errors from local and CI machines resulted in noise.

Just pointing this out because there might be a way to keep the setup leaner by only exporting the token in CI. This would also make things safer. Tracking tokens locally via git-ignored `.env` leaves the door open to them being accidentally uploaded if the file name or location changes. 

Granted, leaking the Sentry API token would not be a big security issue, but once `.env` is in place, more valuable and risky secrets might sneak their way in there.

dcalhoun committed:
build: Remove example environment configuration with secret

Avoid guiding others to accidentally leak secrets.

dcalhoun committed:
build: Remove unnecessary `dotenv` dependency

We do not leverage environment variables outside of the CI environment,
which manages environment variables in a different manner.

dcalhoun commented:
> @dcalhoun is there an expectations on local builds to upload source maps?
> 
> The fact that the testing instructions and the code expect a local `.env` makes me think there might be, but I'm not sure whether that's _necessary_. In my experience on other platforms, tracking errors from local and CI machines resulted in noise.

No, that is not an expectation. I agree doing so is unnecessary noise, and would likely cause confusion. 

The `.env` `.git-ignore` configuration preexisted. The testing instructions were an attempt to make verifying the proposed changes easier. 

> [...] Tracking tokens locally via git-ignored `.env` leaves the door open to them being accidentally uploaded if the file name or location changes.

While I do not think leveraging `dotenv` is an inherently bad thing (not implying you do), I can understand an argument that adding `.env.example` with an empty secret assignment makes it easier/guides someone to accidentally leak secrets. We also currently have no other environment variables for `dotenv` to load. I'll remove the example file and `dotenv` dependency given the expressed concerns. 

Thanks for the feedback! 🙇🏻

mokagio commented:
Thanks @dcalhoun ! Appreciate you taking the time to consider and implement my suggestion 🙇‍♂️
